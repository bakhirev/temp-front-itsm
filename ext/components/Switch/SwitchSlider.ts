import styled, { css } from 'styled-components';
import { size as Size } from 'polished';

import { DEFAULT_THEME } from '../common';

const BORDER_RADIUS_BIG = '10px';
const BORDER_RADIUS_SMALL = '12px';

const INNER_CIRCLE_BORDER_WIDTH = '2px';
const INNER_CIRCLE_SIZE_BIG = '20px';
const INNER_CIRCLE_SIZE_SMALL = '16px';

const LABEL_MARGIN = '8px';

const OUTER_CIRCLE_LEFT_BIG = '14px';
const OUTER_CIRCLE_LEFT_SMALL = '10px';

const OUTER_CIRCLE_SIZE_BIG = '48px';
const OUTER_CIRCLE_SIZE_SMALL = '36px';

const OUTER_CIRCLE_TOP_BIG = '14px';
const OUTER_CIRCLE_TOP_SMALL = '10px';

const SWITCH_HEIGHT_BIG = '20px';
const SWITCH_HEIGHT_SMALL = '16px';

const SWITCH_WIDTH_BIG = '36px';
const SWITCH_WIDTH_SMALL = '28px';

export type LabelPosition = 'right' | 'left';
export type Size = 'big' | 'small';

interface ISwitchSliderProps {
  checked: boolean;
  disabled: boolean;
  labelPosition: LabelPosition;
  size: Size;
}

export const SwitchSlider = styled.div<ISwitchSliderProps>`
  object-fit: contain;
  position: relative;
  box-sizing: content-box !important;
  transition: 0.2s;

  ${({ size, labelPosition }) => css`
    ${size === 'small'
      ? Size(SWITCH_HEIGHT_SMALL, SWITCH_WIDTH_SMALL)
      : Size(SWITCH_HEIGHT_BIG, SWITCH_WIDTH_BIG)}
    border-radius: ${size === 'small' ? BORDER_RADIUS_SMALL : BORDER_RADIUS_BIG};
    ${labelPosition === 'right'
      ? `margin-right: ${LABEL_MARGIN};`
      : `margin-left: ${LABEL_MARGIN};`}
  `}

  ${({ checked, disabled, theme }) =>
    disabled
      ? css`
          background-color: ${theme.color.neutral[30]};
        `
      : css`
          background-color: ${checked ? theme.color.primary[60] : theme.color.neutral[50]};
          &:hover {
            background-color: ${checked ? theme.color.primary[70] : theme.color.neutral[60]};
            &:before {
              border-color: ${checked ? theme.color.primary[70] : theme.color.neutral[60]};
            }
          }
        `};

  &:before {
    content: '';
    position: relative;
    display: inline-block;
    transition: 0.2s;
    box-sizing: border-box;

    ${({ size, theme }) => css`
      ${size === 'small' ? Size(INNER_CIRCLE_SIZE_SMALL) : Size(INNER_CIRCLE_SIZE_BIG)}
      background-color: ${theme.color.neutral.white};
    `}
    border-radius: 50%;
    border: ${INNER_CIRCLE_BORDER_WIDTH} solid
      ${({ checked, disabled, theme }) => {
        if (disabled) {
          return theme.color.neutral[30];
        }
        return checked ? theme.color.primary[60] : theme.color.neutral[50];
      }};
  }

  &:after {
    content: '';
    position: absolute;
    transition: 0.2s;
    box-sizing: border-box;
    ${({ size }) => css`
      ${size === 'small' ? Size(OUTER_CIRCLE_SIZE_SMALL) : Size(OUTER_CIRCLE_SIZE_BIG)}
      top: -${size === 'small' ? OUTER_CIRCLE_TOP_SMALL : OUTER_CIRCLE_TOP_BIG};
      left: -${size === 'small' ? OUTER_CIRCLE_LEFT_SMALL : OUTER_CIRCLE_LEFT_BIG};
    `}
    border-radius: 50%;
    background-color: transparent;
  }
`;

SwitchSlider.defaultProps = {
  theme: DEFAULT_THEME,
};
