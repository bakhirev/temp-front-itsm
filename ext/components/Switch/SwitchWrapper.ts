import styled from 'styled-components';

export const SwitchWrapper = styled.div`
  display: flex;
  align-items: center;
`;
