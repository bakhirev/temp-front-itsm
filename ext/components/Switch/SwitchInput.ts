import styled from 'styled-components';

import { SwitchSlider } from './SwitchSlider';
import type { Size } from './SwitchSlider';

const CIRCLE_TRANSLATE_X_BIG = '16px';
const CIRCLE_TRANSLATE_X_SMALL = '12px';

interface ISwitchInputProps {
  $size: Size;
}

export const SwitchInput = styled.input<ISwitchInputProps>`
  display: none;
  &:checked + ${SwitchSlider} {
    &:before,
    &:after {
      transform: ${({ $size }) =>
        $size === 'small'
          ? `translateX(${CIRCLE_TRANSLATE_X_SMALL});`
          : `translateX(${CIRCLE_TRANSLATE_X_BIG});`};
    }
  }
`;
