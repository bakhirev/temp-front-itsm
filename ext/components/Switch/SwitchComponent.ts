import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../common';
import type { ITheme } from '../themes';

import { SwitchWrapper } from './SwitchWrapper';
import { SwitchSlider } from './SwitchSlider';

interface ISwitchComponentProps {
  disabled: boolean;
}

const commonSwitchComponentStyles = (
  theme: ITheme,
  type: 'Focus' | 'Hover' | 'Pressed',
  disabled: boolean
) => css`
  &:after {
    background-color: ${disabled ? 'transparent' : theme.color.opacity['black' + type]};
  }
`;

export const SwitchComponent = styled.div<ISwitchComponentProps>`
  display: flex;
  align-items: center;
  ${({ disabled, theme }) => css`
    cursor: ${disabled ? 'default' : 'pointer'};
    color: ${disabled ? theme.color.neutral[30] : theme.color.neutral[90]};
  `}
  -webkit-tap-highlight-color: transparent;
  user-select: none;

  &:focus {
    outline: none;
  }

  ${({ theme, disabled }) => css`
    &:focus > ${SwitchWrapper} > ${SwitchSlider} {
      ${commonSwitchComponentStyles(theme, 'Focus', disabled)}
    }
    &:hover > ${SwitchWrapper} > ${SwitchSlider} {
      ${commonSwitchComponentStyles(theme, 'Hover', disabled)}
    }
    &:active > ${SwitchWrapper} > ${SwitchSlider} {
      ${commonSwitchComponentStyles(theme, 'Pressed', disabled)}
    }
  `}
`;

SwitchComponent.defaultProps = {
  theme: DEFAULT_THEME,
};
