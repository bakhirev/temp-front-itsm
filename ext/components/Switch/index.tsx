import React, { FC, MouseEvent, KeyboardEvent } from 'react';

import { Body2Short, Body1Short } from '../Typography';
import { KEY_CODES } from '../common';

import { SwitchComponent } from './SwitchComponent';
import { SwitchInput } from './SwitchInput';
import { SwitchSlider } from './SwitchSlider';
import { SwitchWrapper } from './SwitchWrapper';
import type { LabelPosition, Size } from './SwitchSlider';

export interface ISwitchProps {
  /** Состояние свитча: выставлен/не выставлен */
  checked?: boolean;
  /** Коллбэк на изменение состояния. Срабатывает при клике или переходу по табу */
  onChange: (
    event: MouseEvent<HTMLDivElement> | KeyboardEvent<HTMLDivElement>,
    checked: boolean
  ) => void;
  /** Подпись к свитчу */
  label?: string;
  /** Расположение подписи */
  labelPosition?: LabelPosition;
  /** Отключение свитча */
  disabled?: boolean;
  /** Размер свитча */
  size?: Size;
  /** Имя класса для переопределения стилей */
  className?: string;
  /** data-test-id атрибут для тестирования компонента */
  dataTestId?: string;
}

export const Switch: FC<ISwitchProps> = ({
  label,
  labelPosition = 'right',
  disabled = false,
  size = 'big',
  checked = false,
  className,
  dataTestId,
  onChange,
}) => {
  const handleClick = (event: MouseEvent<HTMLDivElement>) => {
    if (!disabled) {
      onChange && onChange(event, !checked);
    }
  };

  const handleKeyUp = (event: KeyboardEvent<HTMLDivElement>) => {
    if (!disabled && (event.keyCode === KEY_CODES.ENTER || event.keyCode === KEY_CODES.SPACEBAR)) {
      onChange && onChange(event, !checked);
    }
  };

  // To prevent focus event, firing on click
  const handleMouseDown = (event: MouseEvent<HTMLDivElement>) => event.preventDefault();

  const SwitchLabel = size === 'small' ? Body2Short : Body1Short;

  return (
    <SwitchComponent
      className={className}
      data-test-id={dataTestId}
      onKeyUp={handleKeyUp}
      onClick={handleClick}
      onMouseDown={handleMouseDown}
      disabled={disabled}
      tabIndex={1}
    >
      {label && labelPosition === 'left' && <SwitchLabel>{label}</SwitchLabel>}
      <SwitchWrapper>
        <SwitchInput checked={checked} readOnly $size={size} type="checkbox" />
        <SwitchSlider
          checked={checked}
          disabled={disabled}
          labelPosition={labelPosition}
          size={size}
        />
      </SwitchWrapper>
      {label && labelPosition === 'right' && <SwitchLabel>{label}</SwitchLabel>}
    </SwitchComponent>
  );
};
