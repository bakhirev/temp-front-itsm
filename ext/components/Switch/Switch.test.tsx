import React from 'react';
import { shallow } from 'enzyme';

import { Switch } from '../Switch';
import type { ISwitchProps } from '../Switch';

jest.mock('../common', () => ({
  DEFAULT_THEME: {},
}));

describe('Switch', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const switchProps: ISwitchProps = {
    label: 'Switch',
    labelPosition: 'right',
    size: 'small',
    onChange: jest.fn(),
  };

  it('should render component with required properties', () => {
    const wrapper = shallow(<Switch onChange={switchProps.onChange} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with label and small size', () => {
    const wrapper = shallow(<Switch {...switchProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call onClick when user clicks on component', () => {
    const wrapper = shallow(<Switch {...switchProps} />);
    wrapper.simulate('click');
    const { onChange } = switchProps;
    expect(onChange).toBeCalledTimes(1);
  });
});
