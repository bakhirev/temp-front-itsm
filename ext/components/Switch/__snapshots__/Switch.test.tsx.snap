// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch should render component with label and small size 1`] = `
<styled.div
  disabled={false}
  onClick={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  tabIndex={1}
  theme={Object {}}
>
  <styled.div>
    <styled.input
      $size="small"
      checked={false}
      readOnly={true}
      type="checkbox"
    />
    <styled.div
      checked={false}
      disabled={false}
      labelPosition="right"
      size="small"
      theme={Object {}}
    />
  </styled.div>
  <styled.div>
    Switch
  </styled.div>
</styled.div>
`;

exports[`Switch should render component with required properties 1`] = `
<styled.div
  disabled={false}
  onClick={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  tabIndex={1}
  theme={Object {}}
>
  <styled.div>
    <styled.input
      $size="big"
      checked={false}
      readOnly={true}
      type="checkbox"
    />
    <styled.div
      checked={false}
      disabled={false}
      labelPosition="right"
      size="big"
      theme={Object {}}
    />
  </styled.div>
</styled.div>
`;
